kernel/proc.c

```diff
@@ -89,6 +89,22 @@ myproc(void)
   return p;
 }
 
+// Return the struct proc * by pid, or zero if none.
+static struct proc*
+proc_by_pid(int pid)
+{ 
+  struct proc *caller_proc = myproc();
+  struct proc * p;
+
+  for(p = proc; p < &caller_proc[NPROC]; p++) {
+      if (p->pid == pid && !proc->killed) {
```

Comment:
race condition is here

Reply:
wtf?! It should not work!

Reply:
Here, caller_proc -- myproc, not an array,
proc is an array. Also proc->killed is an mistake p->killed is correct